slave-new/process-voms/lib/process-voms.pl

```diff
@@ -0,0 +1,255 @@
+#!/usr/bin/perl
+
+# initialize parser and read the file
+use Sys::Syslog;
+use XML::Simple;
+use Text::CSV;
+use Data::Dumper;
+use Array::Utils qw(:all);
+use JSON::XS;
+my $vos = XMLin( '-',
+	ForceArray => [ 'role', 'group', 'user', 'vo' ],
+#	GroupTags => { role => 'roles', groups => 'group', users => 'user', vos => 'vo' },
+	KeyAttr => [] );
+my $csv = Text::CSV->new({ sep_char => ',' });
+
+### serialize is used to turn an array of hash references into a manageable structure
+sub serialize {
+    JSON::XS->new->relaxed(0)->ascii(1)->canonical(1)->encode($_[0]);
+}
+
+### array_minus_deep is a replacement for array_minu that expands hash references
+sub array_minus_deep(\@\@) {
+    my ($array,$minus) = @_;
+
+    my %minus = map( ( serialize($_) => 1 ), @$minus );
+    grep !$minus{ serialize($_) }, @$array
+}
+
+### getCN extracts a CN from DN. It accepts one aregument:
+#	DN	DN to process
+sub getCN {
+	my $cn = shift;
+	$cn =~ s/.*\/CN=//;
+	$cn =~ s/\/.*//;
+	return $cn;
+}
+
+### listToHashes accepts a three-column CSV and produces an array of hashes with the following structure:
+#	DN	VO Member DN
+#	CA	Certificate Authority that vouches for the member
+#	CN	VO Member CD, extracted from DN
+#	email	The email address of the user
+sub listToHashes {
+	my @hashes;
+	foreach $line (@_) {
+		chomp($line);
+		$csv->parse($line);
+		my @components = $csv->fields();
+		if ( scalar @components > 1 ) { #Crude way to filter out "No members..." messages
+			my %mbr= ( 'DN' => "${components[0]}",'CA' => "${components[1]}", 'CN' => getCN($components[0]), 'email' => "${components[2]}" );
+			push( @hashes, \%mbr );
+		}
+	}
+	return \@hashes;
+}
+
+
+### effectCall runs actual voms-admin commands. It accepts three arguments:
+#	$command	The shell command to run
+#	$debugMsg	Message to log on execution
+sub effectCall {
+	$command = shift;
+	$debugMsg = shift;
+
+#	printf "$command\n\n";
+	@out=`$command 2>&1`;
+
+	if ( $? == 0 ) {
+		syslog(LOG_INFO, "Done $debugMsg");
+	}
+	else {
+		chomp(@out);
+		syslog(LOG_ERR, "Failed $debugMsg Original message: @out");
+		print STDERR "Failed $debugMsg\nOriginal message: @out\nOriginal command: $command\n";
+	}
+}
+
+### knownCA indicates whether a given CA is known to the VOMS server. It accepts the user structure
+#	%user		The user whose CA should be checked (DN, CA, email)
+#	%list		Reference to the list of known CAs
+sub knownCA {
+	$ca = shift;
+	$list = shift;
+
+        if( "$ca" ~~ @{$list}) {
```

Comment:
Be careful about smartmatch, it is deprecated and experimental operator in higher version of Perl so maybe something like:
no if $] >= 5.018, warnings => "experimental";
will be needed.